Remora.Discord.Hosting/Services/DiscordService.cs

```diff
@@ -0,0 +1,77 @@
+//
+//  DiscordService.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) 2017 Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Remora.Discord.Gateway;
+
+namespace Remora.Discord.Hosting.Services
+{
+    /// <summary>
+    /// The <see cref="IHostedService"/> that will run discord in the background.
+    /// </summary>
+    public class DiscordService : BackgroundService
+    {
+        private readonly DiscordGatewayClient _gatewayClient;
+        private readonly ILogger<DiscordService> _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DiscordService"/> class.
+        /// </summary>
+        /// <param name="gatewayClient">The <see cref="DiscordGatewayClient"/>.</param>
+        /// <param name="logger">The <see cref="ILogger"/>.</param>
+        public DiscordService(DiscordGatewayClient gatewayClient, ILogger<DiscordService> logger)
+        {
+            this._gatewayClient = gatewayClient;
+            this._logger = logger;
+        }
+
+        /// <inheritdoc />
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                var runResult = await this._gatewayClient.RunAsync(stoppingToken);
```

Comment:
Generally, the gateway client's `RunAsync` method shouldn't return unless a fatal, non-recoverable error has occurred, or cancellation has been requested via the token. Keeping it alive in a loop like this should not be required.

When I say non-recoverable, I mean in the class of things like "your token is invalid" or "there's a bug that's throwing and it's going to happen again". There *might* be a case where the internet connection isn't available and because of that it returns, and then a loop like this might be warranted - I'm unsure at the time of writing whether that's implicitly handled somewhere in the inner chain, though. Mental note to check.

Reply:
Should I then make the loop only continue if there is a value in In the Websocket / Gateway close status?

I want to avoid any chance of the bot seemingly stopping for no reason, hence why the 5 second delay and attempting to reconnect.

Reply:
That's quite reasonable :) While it shouldn't be neccesary, it's not a bad precaution.